kvdb-lmdb/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! KeyValueDB implementation backed by [LMDB](http://www.lmdb.tech/doc/).
+//!
+//! ## Memory usage
+//!
+//! RSS measurement may report a process as having an entire database resident,
+//! but don't be alarmed. LMDB uses pages in file-backed memory maps (OS page cache),
+//! which can be reclaimed by the OS at any moment
+//! as long as the pages in the map are clean.
+//!
+//! ## Limitations
+//!
+//! - Max key size is 511 bytes (compile time constant `MDB_MAXKEYSIZE`).
+//! - Max threads performing read-only transactions default to 126.
+//! - There is only one active writer allowed at any point of time,
+//! other writers will be blocked until that active writer aborts/commits.
+
+
+#![deny(missing_docs)]
+
+use std::ops::{Deref, DerefMut};
+use std::path::Path;
+use std::io;
+
+use kvdb::{
+	DBValue, NumColumns,
+	TransactionHandler, IterationHandler, OpenHandler,
+	ReadTransaction, WriteTransaction,
+};
+use lmdb::{
+	Environment, Database as LmdbDatabase, DatabaseFlags,
+	Transaction, RoTransaction, RwTransaction,
+	Iter as LmdbIter, Cursor, RoCursor,
+	Error, WriteFlags,
+};
+
+use owning_ref::OwningHandle;
+
+pub use kvdb::DatabaseWithCache;
+
+fn other_io_err<E>(e: E) -> io::Error where E: ToString {
+	io::Error::new(io::ErrorKind::Other, e.to_string())
+}
+
+type KeyValuePair = (Box<[u8]>, Box<[u8]>);
+
+/// Lmdb config.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct DatabaseConfig {
+	num_columns: u32,
+}
+
+impl NumColumns for DatabaseConfig {
+	fn num_columns(&self) -> usize {
+		self.num_columns as usize
+	}
+}
+
+impl OpenHandler<EnvironmentWithDatabases> for EnvironmentWithDatabases {
+	type Config = DatabaseConfig;
+
+	fn open(config: &Self::Config, path: &str) -> io::Result<Self> {
+		Self::open(&Path::new(path), config.num_columns)
+	}
+}
+
+impl<'a> IterationHandler for &'a EnvironmentWithDatabases {
+	type Iterator = IterWithTxn<'a>;
+
+	fn iter(&self, col: usize) -> Self::Iterator {
+		// TODO: how to handle errors properly?
+		let ro_txn = self.ro_txn().expect("lmdb: transaction creation failed");
```

Comment:
Read-only transaction creation can happen if too many thread are trying to create a read-only transaction or we're out of memory (cf. http://www.lmdb.tech/doc/group__mdb.html#gad7ea55da06b77513609efebd44b26920)